pex/glibc.py

```diff
@@ -0,0 +1,82 @@
+# This file was copied from the pip project master branch on 2016/12/05
+from __future__ import absolute_import
+
+import ctypes
+import platform
+import re
+import warnings
+
+
+def glibc_version_string():
+  "Returns glibc version string, or None if not using glibc."
+
+  # ctypes.CDLL(None) internally calls dlopen(NULL), and as the dlopen
+  # manpage says, "If filename is NULL, then the returned handle is for the
+  # main program". This way we can let the linker do the work to figure out
+  # which libc our process is actually using.
+  process_namespace = ctypes.CDLL(None)
+  try:
+    gnu_get_libc_version = process_namespace.gnu_get_libc_version
+  except AttributeError:
+    # Symbol doesn't exist -> therefore, we are not linked to
+    # glibc.
+    return None
+
+  # Call gnu_get_libc_version, which returns a string like "2.5"
+  gnu_get_libc_version.restype = ctypes.c_char_p
+  version_str = gnu_get_libc_version()
+  # py2 / py3 compatibility:
+  if not isinstance(version_str, str):
+    version_str = version_str.decode("ascii")
+
+  return version_str
+
+
+# Separated out from have_compatible_glibc for easier unit testing
+def check_glibc_version(version_str, required_major, minimum_minor):
+  # Parse string and check against requested version.
+  #
+  # We use a regexp instead of str.split because we want to discard any
+  # random junk that might come after the minor version -- this might happen
+  # in patched/forked versions of glibc (e.g. Linaro's version of glibc
+  # uses version strings like "2.20-2014.11"). See gh-3588.
+  m = re.match(r"(?P<major>[0-9]+)\.(?P<minor>[0-9]+)", version_str)
```

Comment:
Does this run frequently enough to be worth pre-compiling in a static?

Reply:
don't believe it's necessary - this runs roughly once per platform x interpreter intersection, so e.g. 4x for an osx+linux x 2.x+3.x pex build.